File: nakshatra_sound.py
"""
nakshatra_sound.py - Pure sound architecture for 27 Nakshatras
No astrology, only music & mantra attributes for 49D kernel integration
"""

import json
import os
from typing import Dict, Optional

class NakshatraSound:
    """Music & mantra mapping for each nakshatra"""
    
    def __init__(self, json_path: str = None):
        if json_path is None:
            json_path = os.path.join(os.path.dirname(__file__), "data", "nakshatra_sound.json")
        
        with open(json_path, "r", encoding="utf-8") as f:
            self.data = json.load(f)
        
        # Create lookup by name and index
        self.by_name: Dict[str, dict] = {}
        self.by_index: Dict[int, dict] = {}
        
        for n in self.data["nakshatras"]:
            self.by_name[n["name"].lower()] = n
            self.by_index[n["index"]] = n
        
        print(f"✅ Loaded {len(self.by_name)} nakshatra sound mappings")
    
    def get_by_name(self, name: str) -> Optional[dict]:
        """Get nakshatra sound data by name (case-insensitive)"""
        return self.by_name.get(name.lower())
    
    def get_by_index(self, idx: int) -> Optional[dict]:
        """Get nakshatra sound data by index (1-27)"""
        return self.by_index.get(idx)
    
    def get_svara_frequency(self, svara: str, base_freq: float = 261.63) -> float:
        """
        Convert svara name to frequency ratio
        Base frequency: C4 (Middle C) = 261.63 Hz
        """
        ratios = {
            "Sa (śuddha)": 1/1,
            "Re (komal)": 9/8,
            "Ga (śuddha)": 5/4,
            "Ma (tivra)": 4/3,
            "Pa (śuddha)": 3/2,
            "Dha (komal)": 11/8,
            "Ni (komal)": 7/4
        }
        ratio = ratios.get(svara, 1/1)
        return base_freq * ratio
    
    def get_meter_pattern(self, chand: str) -> dict:
        """
        Convert meter name to syllable pattern
        Returns pattern and syllable count
        """
        meters = {
            "Vasantatilaka": {"pattern": "S S G S G G", "syllables": 12},
            "Shikharini": {"pattern": "G G G S G S G G G", "syllables": 17},
            "Mandākrāntā": {"pattern": "G S G S G S G S G", "syllables": 15},
            "Upajati": {"pattern": "G G S G G G", "syllables": 11},
            "Shārdūlavikrīḍita": {"pattern": "G G S G G S G G G", "syllables": 19}
        }
        return meters.get(chand, {"pattern": "S G", "syllables": 2})
    
    def to_49d_vector(self, nakshatra: dict) -> list:
        """
        Convert nakshatra sound attributes to 49-dimensional vector
        Each attribute maps to specific dimension ranges
        """
        vec = [0.0] * 49
        
        # Svara (dimensions 1-7) - one-hot based on note
        svara_map = {
            "Sa (śuddha)": 0, "Re (komal)": 1, "Ga (śuddha)": 2,
            "Ma (tivra)": 3, "Pa (śuddha)": 4, "Dha (komal)": 5,
            "Ni (komal)": 6
        }
        svara_idx = svara_map.get(nakshatra["svara"], 0)
        vec[svara_idx] = 1.0
        
        # Raga (dimensions 8-20) - encode as frequency cluster
        raga_freq = hash(nakshatra["raga"]) % 13
        vec[8 + raga_freq] = 0.8
        
        # Chand/meter (dimensions 21-28) - syllable count
        meter = self.get_meter_pattern(nakshatra["chand"])
        vec[21] = meter["syllables"] / 20.0
        
        # Bija phoneme (dimensions 29-40) - phonetic class
        bija_val = ord(nakshatra["bija"][0]) % 12
        vec[29 + bija_val] = 0.9
        
        # Lyric seed (dimensions 41-49) - semantic anchor
        seed_val = len(nakshatra["seed"]) / 10.0
        vec[41] = seed_val
        
        return vec
    
    def generate_mantra_phrase(self, nakshatra: dict) -> str:
        """Generate a simple mantra phrase using bija and seed"""
        bija = nakshatra["bija"].lower()
        seed = nakshatra["seed"].lower()
        
        if len(bija) == 1:
            return f"Oṁ {bija}ṁ {seed} namaḥ"
        else:
            return f"Oṁ {bija} {seed} namaḥ"

# Singleton instance
_sound_db = None

def get_sound_db() -> NakshatraSound:
    global _sound_db
    if _sound_db is None:
        _sound_db = NakshatraSound()
    return _sound_db

# Quick test
if __name__ == "__main__":
    db = get_sound_db()
    rohini = db.get_by_name("Rohiṇī")
    print(f"\n📊 Rohiṇī Sound Data:")
    print(f"   Svara: {rohini['svara']}")
    print(f"   Rāga: {rohini['raga']}")
    print(f"   Chand: {rohini['chand']}")
    print(f"   Bīja: {rohini['bija']}")
    print(f"   Seed: {rohini['seed']}")
    print(f"   Mantra: {db.generate_mantra_phrase(rohini)}")
    print(f"   Frequency: {db.get_svara_frequency(rohini['svara']):.1f} Hz")
    
    # Test 49D vector
    vec = db.to_49d_vector(rohini)
    print(f"\n   49D Vector (first 10 dims): {vec[:10]}")

File: nakshatra_sound.js
/**
 * nakshatra_sound.js - Pure sound architecture for 27 Nakshatras
 * No astrology, only music & mantra attributes
 */

const NAKSHATRA_SOUND_DATA = {
  "version": "1.0",
  "nakshatras": [
    {"index":1,"name":"Ashwinī","svara":"Ni (komal)","raga":"Bhairav","chand":"Vasantatilaka","bija":"A","seed":"Aśva"},
    {"index":2,"name":"Bharaṇī","svara":"Ma (tivra)","raga":"Todi","chand":"Shikharini","bija":"Ī","seed":"Īśā"},
    {"index":3,"name":"Kṛttikā","svara":"Ga (śuddha)","raga":"Marwa","chand":"Mandākrāntā","bija":"U","seed":"Ulkā"},
    {"index":4,"name":"Rohiṇī","svara":"Re (komal)","raga":"Khamaj","chand":"Vasantatilaka","bija":"AI","seed":"Ena"},
    {"index":5,"name":"Mrigaśirā","svara":"Sa (śuddha)","raga":"Bageshree","chand":"Upajati","bija":"Ē","seed":"Mriga"},
    {"index":6,"name":"Ārdrā","svara":"Dha (komal)","raga":"Malkauns","chand":"Shārdūlavikrīḍita","bija":"O","seed":"Rudra"},
    {"index":7,"name":"Punarvasu","svara":"Pa (śuddha)","raga":"Yaman","chand":"Vasantatilaka","bija":"AU","seed":"Aditi"},
    {"index":8,"name":"Puṣya","svara":"Ni (komal)","raga":"Darbari","chand":"Upajati","bija":"AM","seed":"Brhaspati"},
    {"index":9,"name":"Āśleṣā","svara":"Ga (śuddha)","raga":"Bhimpalasi","chand":"Mandākrāntā","bija":"AH","seed":"Nāga"},
    {"index":10,"name":"Maghā","svara":"Re (komal)","raga":"Bhairav","chand":"Shikharini","bija":"KA","seed":"Pitṛ"},
    {"index":11,"name":"Pūrva Phālgunī","svara":"Ma (tivra)","raga":"Khamaj","chand":"Vasantatilaka","bija":"KHA","seed":"Bhaga"},
    {"index":12,"name":"Uttara Phālgunī","svara":"Sa (śuddha)","raga":"Bhoop","chand":"Upajati","bija":"GA","seed":"Aryaman"},
    {"index":13,"name":"Hasta","svara":"Dha (komal)","raga":"Yaman","chand":"Mandākrāntā","bija":"GHA","seed":"Savitṛ"},
    {"index":14,"name":"Citrā","svara":"Ni (komal)","raga":"Brindavani Sarang","chand":"Shikharini","bija":"CA","seed":"Viśva"},
    {"index":15,"name":"Svāti","svara":"Pa (śuddha)","raga":"Hamsadhwani","chand":"Vasantatilaka","bija":"CHA","seed":"Vāyu"},
    {"index":16,"name":"Viśākhā","svara":"Ga (śuddha)","raga":"Shankara","chand":"Upajati","bija":"JA","seed":"Indra"},
    {"index":17,"name":"Anurādhā","svara":"Re (komal)","raga":"Malkauns","chand":"Mandākrāntā","bija":"JHA","seed":"Mitra"},
    {"index":18,"name":"Jyeṣṭhā","svara":"Ma (tivra)","raga":"Darbari","chand":"Shikharini","bija":"ÑA","seed":"Indra"},
    {"index":19,"name":"Mūla","svara":"Sa (śuddha)","raga":"Malkauns","chand":"Vasantatilaka","bija":"ṬA","seed":"Nirṛti"},
    {"index":20,"name":"Pūrva Aṣāḍhā","svara":"Dha (komal)","raga":"Khamaj","chand":"Upajati","bija":"ṬHA","seed":"Apa"},
    {"index":21,"name":"Uttara Aṣāḍhā","svara":"Ni (komal)","raga":"Bhoop","chand":"Mandākrāntā","bija":"ḌA","seed":"Viśve"},
    {"index":22,"name":"Śravaṇa","svara":"Pa (śuddha)","raga":"Yaman","chand":"Shikharini","bija":"ḌHA","seed":"Viṣṇu"},
    {"index":23,"name":"Dhanisṭhā","svara":"Ga (śuddha)","raga":"Brindavani Sarang","chand":"Vasantatilaka","bija":"ṆA","seed":"Vasu"},
    {"index":24,"name":"Śatabhiṣā","svara":"Re (komal)","raga":"Bhairav","chand":"Upajati","bija":"TA","seed":"Varuṇa"},
    {"index":25,"name":"Pūrva Bhādrapadā","svara":"Ma (tivra)","raga":"Desh","chand":"Mandākrāntā","bija":"THA","seed":"Aja"},
    {"index":26,"name":"Uttara Bhādrapadā","svara":"Sa (śuddha)","raga":"Darbari","chand":"Shikharini","bija":"DA","seed":"Ahir"},
    {"index":27,"name":"Revatī","svara":"Dha (komal)","raga":"Bhimpalasi","chand":"Vasantatilaka","bija":"DHA","seed":"Pūṣan"}
  ]
};

class NakshatraSoundJS {
  constructor() {
    this.data = NAKSHATRA_SOUND_DATA;
    this.byName = {};
    this.byIndex = {};
    
    for (const n of this.data.nakshatras) {
      this.byName[n.name.toLowerCase()] = n;
      this.byIndex[n.index] = n;
    }
  }
  
  getByName(name) {
    return this.byName[name.toLowerCase()];
  }
  
  getByIndex(idx) {
    return this.byIndex[idx];
  }
  
  getSvaraFrequency(svara, baseFreq = 261.63) {
    const ratios = {
      "Sa (śuddha)": 1/1,
      "Re (komal)": 9/8,
      "Ga (śuddha)": 5/4,
      "Ma (tivra)": 4/3,
      "Pa (śuddha)": 3/2,
      "Dha (komal)": 11/8,
      "Ni (komal)": 7/4
    };
    const ratio = ratios[svara] || 1;
    return baseFreq * ratio;
  }
  
  generateMantra(nakshatra) {
    const bija = nakshatra.bija.toLowerCase();
    const seed = nakshatra.seed.toLowerCase();
    return `Oṁ ${bija}ṁ ${seed} namaḥ`;
  }
}

// Export for browser
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { NakshatraSoundJS, NAKSHATRA_SOUND_DATA };
}
